src/BenchmarkDotNet/Toolchains/DotNetSdkProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace BenchmarkDotNet.Toolchains
+{
+    public class DotNetSdkProvider : IDotNetSdkProvider
+    {
+        private readonly string _customDotNetCliPath;
+
+        public string CustomDotNetCliPath => _customDotNetCliPath;
+
+        public DotNetSdkProvider(string customDotNetCliPath = null)
+        {
+            _customDotNetCliPath = customDotNetCliPath;
+        }
+
+        public IEnumerable<string> GetInstalledSdks()
+        {
+            var installedDotNetSdks = GetInstalledDotNetSdk();
+            var installedFrameworkSdks = GetInstalledFrameworkSdks();
+            return installedDotNetSdks.Concat(installedFrameworkSdks);
+        }
+
+        private IEnumerable<string> GetInstalledDotNetSdk()
+        {
+            string dotnetExecutable = string.IsNullOrEmpty(CustomDotNetCliPath) ? "dotnet" : CustomDotNetCliPath;
+            var startInfo = new ProcessStartInfo(dotnetExecutable, "--list-sdks")
+            {
+                RedirectStandardOutput = true,
+                UseShellExecute = false,
+                CreateNoWindow = true
+            };
+
+            using (var process = Process.Start(startInfo))
+            {
+                if (process == null) throw new InvalidOperationException("Failed to start the dotnet process.");
```

Comment:
I think this should return empty instead of throw. You can check `HostEnvironmentInfo.GetCurrent().IsDotNetCliInstalled()` before running the process.

Reply:
Ok I just committed some changes with separated .Net and .Net Framework providers. Would you want me to cache both results to `HostEnvironmentInfo`?

Reply:
I guess it doesn't make sense to cache the core versions in `HostEnvironmentInfo` if a custom cli can be used. I'm not sure how expensive retrieving the Framework sdks is, but it would probably be good to cache the results in case of multiple jobs using the same toolchain. The cache could live in the `DotNetSdkVersionValidator` rather than `HostEnvironmentInfo` I guess (and you could cache core sdks per custom cli if you think it's worth it).